### PR TITLE
Added inline comments for config files

### PR DESCRIPTION
**Pull request by anushka255**
The comments for `config` and `config_fast` are exactly the same.

### CODE REVIEW

**Comment by skinnider** [comment]
Awesome, thanks @anushka255 !
@spavelites, you are probably the best judge of whether the inline comments for each parameter in the workflow are clear or any need clarification - any feedback?

**Comment by vineetbansal** [comment]
@anushka255 - I don't think this is true? Each train seed applies to each input fold afaik.
```
The number of 'train_seeds' have to the same as the number of folds specified in line 8.
```

File: workflow/config/config.yaml [hunk 1]
```diff
@@ -12,69 +21,108 @@ train_seeds:
...
+
+# Seeds used when sampling molecules from the trained models.
+# The amount of 'sample_seeds' parameter specifies how many times the 'sample_molecules_RNN' step is executed,
+# each time using the same trained model but with different random seed values.
 sample_seeds:
   - 0
+
+# Specifies by how many times the input data is augmented (or enumerated) before training.
+# This utilizes the fact that a single molecule can have multiple SMILES representation.
+# For example, a value of 0 means no augmentation, while 100 means each molecule can have
```
Comment (vineetbansal) [inline workflow/config/config.yaml:33]: might be nice to explain how '0' is different from '1'. Also, is it "up to 100" or "exactly 100"? Might be worthwhile looking at the implementation.

File: workflow/config/config.yaml [hunk 2]
```diff
@@ -12,69 +21,108 @@ train_seeds:
...
+# up to 100 different representations in the training set.
 enum_factors:
   - 0
   - 10
   - 30
   - 50
   - 100
+
+#  Limits the maximum number of input SMILES to be processed. 0 means there's no limit.
```
Comment (vineetbansal) [inline workflow/config/config.yaml:42]: in the processing step? I forgot if this limit applies to more than one step.

**Comment by vineetbansal** [comment]
Thanks @anushka255 - minor gripe - since the config lines can get arbitrarily long (especially the paths), I'd prefer the comments to go above the relevant line instead of after it, so the purpose can be gleaned independent of the value. We might also put in a  yaml formatter in our pre-commit so the formatting is standardized. (it might be there already). The latter is not that important and can be handled as a separate issue I guess.

File: workflow/config/config.yaml [hunk 3]
```diff
@@ -12,69 +21,108 @@ train_seeds:
...
-  n_layers: 3
-  dropout: 0
-  batch_size: 64
-  learning_rate: 0.001
-  max_epochs: 99999
-  patience: 50000
-  log_every_steps: 100
-  log_every_epochs: 1
-  sample_mols: 10000000
+  rnn_type: LSTM # Type of Recurrent Neural Network (RNN) to use.
```
[the remark this answers is not in the document]
Comment (skinnider) [reply workflow/config/config.yaml:47]: Agree - similar for representation (SMILES vs SELFIES being the two options)

[the remark this answers is not in the document]
Comment (vineetbansal) [reply workflow/config/config.yaml:47]: @skinnider - since you raise this as an example - I'm fairly certain that running the workflow on SELFIES will not work at all and will open up a can of worms (I think we realized early on that we needed to handle selfies as a special case in certain steps, and decided to not handle it). Perhaps we remove all mentions of SELFIES? (and add an issue to look into them if we need to?)

[the remark this answers is not in the document]
Comment (skinnider) [reply workflow/config/config.yaml:47]: Yes, that would be very reasonable since I think we will never use SELFIES going forward.

**Comment by anushka255** [comment]
> @anushka255 - I don't think this is true? Each train seed applies to each input fold afaik.
> 
> ```
> The number of 'train_seeds' have to the same as the number of folds specified in line 8.
> ```

@vineetbansal You're right, I'll make the changes.

File: workflow/config/config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (anushka255) [reply workflow/config/config.yaml:47]: I am running a workflow on argo with SELFIES. It's on the training step right now. I'll let you know if anything breaks along the way.

**Comment by anushka255** [comment]
I have made all the changes pointed out in the comments above. If it's okay with you guys we can merge this branch without waiting for the workflows on SELFIES/ GRU to complete. If I find that the results of the workflows are different than what the config files mention, I will open a separate pull request to address that.

**Comment by skinnider** [comment]
Not waiting for SELFIES to run sounds good to me but I'm still hoping to get @spavelites feedback on the clarity of the inline comments before we merge.

**Review by skinnider — APPROVED** [review]

**Comment by vineetbansal** [comment]
thanks @anushka255 for merging `master` to this. @skinnider - I'll add comments for the newly-added conditional rnn config options as another commit shortly - perhaps you can take a final look before we merge.

**Comment by vineetbansal** [comment]
@skinnider - I'm adding comments based on your notes to yourself in another PR (thanks!)

- conditional_emb concatenates the descriptors directly to the token embeddings at each step in the sequence; 
- conditional_emb_l first passes them through a linear layer to obtain embeddings of dimensionality equal to that of the token embeddings
- conditional_dec and conditional_dec_l are analogous, except that the descriptors or their embeddings are concatenated to the output of the RNN layers (prior to the decoder layer)
- conditional_h instantiates the hidden states based on learned transformations of the descriptors (with a single linear layer), as in Kotsias et al.

In looking at the code that is validating these different options before starting anything though, the logic doesn't seem quite correct to me:

```
  # Assert that conditional_emb_l and conditional_emb cannot both be true at the same time
  assert not (
      self.conditional_emb_l and self.conditional_emb and self.conditional_h
  ), f"Both conditional_emb_l and conditional_emb cannot be true at the same time. Got: conditional_emb_l={self.conditional_emb_l}, conditional_emb={self.conditional_emb}"

  # Assert that if conditional_emb_l is true, conditional_emb must be false, and vice versa. Assert the same for conditional_dec_l and conditional_dec
  assert (
      (self.conditional_emb_l != self.conditional_emb)
      or (self.conditional_dec_l != self.conditional_dec)
      or self.conditional_h
  ), f"Expected one of conditional_emb_l or conditional_emb to be true and the other false. This should be followed for conditional_dec_l and conditional_dec. If both these conditions are false then conditional_h should be True. Got: conditional_emb_l={self.conditional_emb_l}, conditional_emb={self.conditional_emb}, conditional_dec_l={self.conditional_dec_l}, conditional_dec={self.conditional_dec}, conditional_h={self.conditional_h}"

  # Assert that conditional_dec_l and conditional_dec cannot both be true at the same time
  assert not (
      self.conditional_dec_l and self.conditional_dec and self.conditional_h
  ), f"Both conditional_dec_l and conditional_dec cannot be true at the same time. Got: conditional_dec_l={self.conditional_dec_l}, conditional_dec={self.conditional_dec}"

```

My understanding is that exactly one of `conditional_emb` and `conditional_emb_l` must be `True`, and exactly one of `conditional_dec` and `conditional_dec_l` must be `True`, and `conditional_h` could be `True` or `False` independent of anything else.

So I'm not sure why `conditional_h` is part of the checks above - one comment above indicates that `If both these conditions are false then conditional_h should be True` (which, even if that's what we want to do, is not being checked correctly).

**Comment by skinnider** [comment]
Sorry for the delayed reply. After thinking about this a bit, I think the logic is as follows:
- `conditional_emb` and `conditional_emb_l` cannot both be True, but can both be False, or one can be true
- `conditional_dec` and `conditional_dec_l` cannot both be True, but can both be False, or one can be true
- `conditional_h` is independent of the above
- if at least one of the above is True, then we are training a conditional language model; if none of the above are True, then the model is not conditional

So, for example, the user could enable all three of `conditional_emb`, `conditional_dec`, and `conditional_h`, or only `conditional_h` (the latter is actually probably the mode we will use most often in the future). 

I would say the checks should be:
- not both `conditional_emb` / `conditional_emb_l` True
- same for `conditional_dec`
- if `conditional` is enabled, then at least one of these five must be True
Does that make sense?

**Comment by skinnider** [comment]
looks good to me, too
